libnss/src/interop.rs

```diff
@@ -1,24 +1,72 @@
-use libc::c_int;
 use std::collections::VecDeque;
 use std::ffi::CString;
+use std::io;
+
+pub trait ToC<C> {
+    unsafe fn to_c(&self, result: *mut C, buffer: &mut CBuffer) -> std::io::Result<()>;
+}
 
 #[allow(dead_code)]
 pub enum NssStatus {
+    TryAgain = -2,
+    Unavail = -1,
+    NotFound = 0,
+    Success = 1,
+    Return = 2,
+}
+
+pub enum Response<R> {
     TryAgain,
     Unavail,
     NotFound,
-    Success,
+    Success(R),
     Return,
 }
 
-impl NssStatus {
-    pub fn to_c(&self) -> c_int {
-        match *self {
-            NssStatus::TryAgain => -2,
-            NssStatus::Unavail => -1,
-            NssStatus::NotFound => 0,
-            NssStatus::Success => 1,
-            NssStatus::Return => 2,
+impl<R> Response<R> {
+    pub fn to_status(&self) -> NssStatus {
+        use NssStatus::*;
+        match self {
+            Self::Success(..) => Success,
+            Self::TryAgain => TryAgain,
+            Self::Unavail => Unavail,
+            Self::NotFound => NotFound,
+            Self::Return => Return,
+        }
+    }
+
+    pub unsafe fn to_c<C>(
+        &self,
+        result: *mut C,
+        buf: *mut libc::c_char,
+        buflen: libc::size_t,
+        errnop: *mut libc::c_int,
+    ) -> NssStatus
+    where
+        R: ToC<C>,
+    {
+        if let Self::Success(entity) = self {
+            let mut buffer = CBuffer::new(buf as *mut libc::c_void, buflen);
+            buffer.clear();
+
+            match entity.to_c(result, &mut buffer) {
+                Ok(()) => {
+                    *errnop = 0;
+                    self.to_status()
+                }
+                Err(e) => match e.raw_os_error() {
+                    Some(e) => {
+                        *errnop = e;
+                        Self::TryAgain.to_status()
```

Comment:
This is also covered by #4 but in this PR, the logic to handle an insufficiently large buffer is now consolidated into a single place here instead of repeated in each macro.

Furthermore, as in my comment above about the distinction between `Result` and `NssStatus`, this shows an example of setting `errnop` to a meaningful value (here, handling `ERANGE`, which is the only code the glibc docs _mandate_ being used for this purpose).